### PR TITLE
Full kmip2.0 support including new Attributes

### PR DESCRIPTION
**Pull request by alikdolg**
## Summary by CodeRabbit

* **New Features**
  * Broader KMIP 2.0 operation support (GetAttributes/GetAttributeList/Add/Modify/Delete Attribute, ReKey, Revoke, Encrypt/Decrypt, Activate) and richer attribute payload normalization.

* **Bug Fixes**
  * More robust encode/decode handling (pointer wrapping, field lookup/seen tracking) and legacy fallback behavior including revocation message/date fixes.

* **Tests**
  * Extensive new unit tests covering attribute edge cases, legacy vs new payloads, and operation encode/decode scenarios.

### CODE REVIEW

File: request.go [hunk 1]
```diff
@@ -72,5 +72,5 @@ type RevocationReason struct {
 	Tag `kmip:"REVOCATION_REASON"`
 
 	RevocationReasonCode Enum   `kmip:"REVOCATION_REASON_CODE"`
-	RevocationMessage    string `kmip:"REVOCATION_REASON"`
+	RevocationMessage    string `kmip:"REVOCATION_MESSAGE"`
```
Comment (tuvia-akeyless) [inline request.go]: BC?

File: payload_registry.go [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package kmip
 
 import (
+	"fmt"
```
Comment (tuvia-akeyless) [inline payload_registry.go]: cleanup prints?

File: attribute.go [hunk 3]
```diff
@@ -18,6 +18,17 @@ type Attribute struct {
...
+	NameValue              Name  `kmip:"NAME"`
+	ObjectType             Enum  `kmip:"OBJECT_TYPE"`
+	CryptographicAlgorithm Enum  `kmip:"CRYPTOGRAPHIC_ALGORITHM"`
+	CryptographicLength    int32 `kmip:"CRYPTOGRAPHIC_LENGTH"`
+	CryptographicUsageMask int32 `kmip:"CRYPTOGRAPHIC_USAGE_MASK"`
+	Sensitive              bool  `kmip:"SENSITIVE"`
+	AlwaysSensitive        bool  `kmip:"ALWAYS_SENSITIVE"`
+	Extractable            bool  `kmip:"EXTRACTABLE"`
+	NeverExtractable       bool  `kmip:"NEVER_EXTRACTABLE"`
+	ReplaceExisting        bool  `kmip:"REPLACE_EXISTING"`
```
Comment (tuvia-akeyless) [inline attribute.go:32]: Consider to add comment that the following fields were added in kmip 2.0

File: decode.go [hunk 4]
```diff
@@ -338,24 +345,58 @@ func (d *Decoder) decode(rv reflect.Value, structD *structDesc) (n int, err erro
...
+			if f.tag == tag {
+				if fallbackFieldIdx == -1 {
+					fallbackFieldIdx = i
+				}
+				if f.typ == typ {
+					fieldIdx = i
+					break
+				}
+			}
+		}
```
Comment (tuvia-akeyless) [inline decode.go:385]: What if you find typ before you find tag - so you break before setting `fallbackFieldIdx` with the tag index. If it is on purpose, please add comment.

File: decode.go [hunk 5]
```diff
@@ -338,24 +345,58 @@ func (d *Decoder) decode(rv reflect.Value, structD *structDesc) (n int, err erro
...
+		if fieldIdx == -1 {
+			fieldIdx = fallbackFieldIdx
+		}
```
Comment (tuvia-akeyless) [inline decode.go:390]: This means that we take the index of the tag, if typ not found? Can add comment why to do it?

File: decode.go [hunk 6]
```diff
@@ -338,24 +345,58 @@ func (d *Decoder) decode(rv reflect.Value, structD *structDesc) (n int, err erro
...
+		if fieldIdx == -1 {
+			for i, f := range structD.fields {
+				if f.tag == ANY_TAG {
+					fieldIdx = i
+					break
+				}
+			}
+		}
```
Comment (tuvia-akeyless) [inline decode.go:400]: What is the meaning of `ANY_TAG`? What if there are few ANY_TAGs in the fields map?

File: decode.go [hunk 7]
```diff
@@ -394,6 +437,11 @@ func (d *Decoder) decode(rv reflect.Value, structD *structDesc) (n int, err erro
 				}
 			}
 		} else {
+			if seen[fieldIdx] {
+				err = errors.Errorf("duplicate field %v", f.name)
+				return
+			}
```
Comment (tuvia-akeyless) [inline decode.go:450]: Can it break BC if clients used duplicate fields mistakely?

File: kmip20/payloads.go [hunk 8]
```diff
@@ -90,13 +130,394 @@ type Attributes struct {
...
+	Extractable            bool      `kmip:"EXTRACTABLE"`
+	NeverExtractable       bool      `kmip:"NEVER_EXTRACTABLE"`
+	ReplaceExisting        bool      `kmip:"REPLACE_EXISTING"`
+
+	AttributeName  string      `kmip:"ATTRIBUTE_NAME"`
+	AttributeIndex int32       `kmip:"ATTRIBUTE_INDEX"`
+	AttributeValue interface{} `kmip:"ATTRIBUTE_VALUE"`
+
+	NewAttributes     []Attributes `kmip:"NEW_ATTRIBUTE"`
+	CurrentAttributes []Attributes `kmip:"CURRENT_ATTRIBUTE"`
```
Comment (tuvia-akeyless) [inline kmip20/payloads.go:155]: I didn't understand how these fields were added now, while this 2.0 struct was here before - the protocol didn't change.

File: kmip20/payloads.go [hunk 9]
```diff
@@ -90,13 +130,394 @@ type Attributes struct {
...
+	for _, attr := range a.NewAttributes {
+		if v := attr.Attribute(); v.Name != "" {
+			a.Values = append(a.Values, v)
+		}
+	}
+	for _, attr := range a.CurrentAttributes {
+		if v := attr.Attribute(); v.Name != "" {
+			a.Values = append(a.Values, v)
+		}
+	}
```
Comment (tuvia-akeyless) [inline kmip20/payloads.go:210]: It only appends? How can we remove an existing attributes?

File: kmip20/payloads.go [hunk 10]
```diff
@@ -90,13 +130,394 @@ type Attributes struct {
...
+	if len(a.Values) == 0 && (seenField(seen, "AttributeName") || a.AttributeName != "") {
```
Comment (tuvia-akeyless) [inline kmip20/payloads.go:211]: Why for this field we check also `if len(a.Values) == 0`?

File: kmip20/payloads.go [hunk 11]
```diff
@@ -90,13 +130,394 @@ type Attributes struct {
...
+	return values
+}
+
+func (a *AttributeValues) Set(attr kmip.Attribute) {
+	a.Values = append(a.Values, attr)
+
+	switch attr.Name {
+	case kmip.ATTRIBUTE_NAME_UNIQUE_IDENTIFIER:
+		if v, ok := attr.Value.(string); ok {
+			a.UniqueIdentifier = &v
```
Comment (tuvia-akeyless) [inline kmip20/payloads.go:304]: Why for every field need to set both `a.Values` and the specific field (`a.UniqueIdentifier`)?
BTW, `func (a *AttributeValues) AfterUnmarshalKMIPWithSeenFields` also do it.

File: kmip20/payloads.go [hunk 12]
```diff
@@ -90,13 +130,394 @@ type Attributes struct {
...
+		}
+		return nil, errors.Errorf("unsupported attribute: %v", name)
+	}
 }
 
 // CreateRequest is the KMIP 2.0 Create request payload.
 type CreateRequest struct {
 	ObjectType             kmip.Enum  `kmip:"OBJECT_TYPE,required"`
-	Attributes             Attributes `kmip:"ATTRIBUTES,required"`
+	Attributes             Attributes `kmip:"ATTRIBUTES"`
```
Comment (tuvia-akeyless) [inline kmip20/payloads.go:525]: Please ensure there is no impact on the param type. I know that in swagger for example, removing `required` makes no-pointer to be pointer.

File: request.go (hunk 1 above)
Comment (alikdolg) [reply request.go]: Yes - BC, but it's ok. By spec it is original tag (REVOCATION_MESSAGE) for both 1.x and 2.0 versions.
And this message is not matter - what really matter is REVOCATION_REASON_CODE.

File: payload_registry.go (hunk 2 above)
Comment (alikdolg) [reply payload_registry.go]: Oh - thanks!!!

File: attribute.go (hunk 3 above)
Comment (alikdolg) [reply attribute.go:32]: Done

File: decode.go (hunk 4 above)
Comment (alikdolg) [reply decode.go:385]: The decoder comments now explain the tag/type fallback behavior that the reviewer asked about. I’m adding similar brief notes around the KMIP 2.0 attribute normalization paths.
Added comment

File: decode.go (hunk 5 above)
Comment (alikdolg) [reply decode.go:390]: Added comment

File: decode.go (hunk 6 above)
Comment (alikdolg) [reply decode.go:400]: `
```
// KMIP Tags
const (
	// Internal
	ANY_TAG Tag = 0xffffff
	// KMIP 1.0
	ACTIVATION_DATE                        Tag = 0x420001
	APPLICATION_DATA                       Tag = 0x420002
	APPLICATION_NAMESPACE                  Tag = 0x420003
	..............
```
	`

Comment (alikdolg) [reply decode.go:400]: ANY_TAG is only used as a wildcard fallback after no concrete tag matches. I found it maps from kmip:"-" / ANY_TAG, so multiple wildcard fields would be ambiguous. If a struct has a few ANY_TAG fields, the decoder will use the first one in structD.fields order and ignore the rest for that incoming tag
Added comment

File: decode.go (hunk 7 above)
Comment (alikdolg) [reply decode.go:450]: Yes, it can break backward compatibility for invalid payloads.
Repeated fields that are valid are represented as slices, so this only affects duplicate scalar fields.
I think failing fast is preferable here because duplicate scalar tags are ambiguous

File: kmip20/payloads.go (hunk 8 above)
Comment (alikdolg) [reply kmip20/payloads.go:155]: It is changes since first merge. I use it for migration from Data model to Attribute

File: kmip20/payloads.go (hunk 9 above)
Comment (alikdolg) [reply kmip20/payloads.go:210]: This hook only normalizes decoded payload shapes into Values; it does not apply attribute mutations. Removal is represented by DeleteAttributeRequest using CURRENT_ATTRIBUTE or ATTRIBUTE_REFERENCE. Appending here preserves the attributes present on the wire so the operation handler can decide what to delete.

File: kmip20/payloads.go (hunk 10 above)
Comment (alikdolg) [reply kmip20/payloads.go:211]: Without len(a.Values) == 0, a payload containing both a nested/direct attribute and a legacy wrapper would add both. That could make Attribute() return the wrong first value or make callers see duplicate attributes.

File: kmip20/payloads.go (hunk 11 above)
Comment (alikdolg) [reply kmip20/payloads.go:304]: Because AttributeValues is doing two jobs:

a.Values is the normalized Go-facing list of attributes. Callers can read attributes uniformly from Values regardless of how they appeared on the wire.
The specific fields like a.UniqueIdentifier are the KMIP 2.0 wire representation. They let the encoder emit direct tags like UNIQUE_IDENTIFIER, SENSITIVE, CRYPTOGRAPHIC_USAGE_MASK, instead of legacy ATTRIBUTE wrappers. They are pointers so explicit zero values such as false or int32(0) are still encoded.

So in Set():

a.Values = append(a.Values, attr)
a.UniqueIdentifier = &v
means:

keep the canonical in-memory list populated
also prepare the direct KMIP 2.0 field for encoding

File: kmip20/payloads.go (hunk 12 above)
Comment (alikdolg) [reply kmip20/payloads.go:525]: Added comment

File: kmip20/payloads.go (hunk 11 above)
Comment (tuvia-akeyless) [reply kmip20/payloads.go:304]: Discussed - Need both datas to work with both kmip versions. Specifically for Attributes.

**Review by tuvia-akeyless — APPROVED** [review]
